ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsMaterializedTable.java

```diff
@@ -60,13 +70,19 @@
   @Override
   public Iterator<Row> get(final int partition) {
     try {
-      final ReadOnlyKeyValueStore<GenericKey, ValueAndTimestamp<GenericRow>> store = stateStore
-          .store(QueryableStoreTypes.timestampedKeyValueStore(), partition);
-
-      final KeyValueIterator<GenericKey, ValueAndTimestamp<GenericRow>> iterator = store.all();
+      final RangeQuery<GenericKey, ValueAndTimestamp<GenericRow>> query = RangeQuery.withNoBounds();
+      final StateQueryRequest<KeyValueIterator<GenericKey, ValueAndTimestamp<GenericRow>>>
+          request = inStore(stateStore.getStateStoreName())
+          .withQuery(query)
+          .withPartitions(ImmutableSet.of(partition));
+      final StateQueryResult<KeyValueIterator<GenericKey, ValueAndTimestamp<GenericRow>>>
+          result = stateStore.getKafkaStreams().query(request);
+      final KeyValueIterator<GenericKey, ValueAndTimestamp<GenericRow>> iterator =
+          result.getOnlyPartitionResult().getResult();
+      
       return Streams.stream(IteratorUtil.onComplete(iterator, iterator::close))
           .map(keyValue -> Row.of(stateStore.schema(), keyValue.key, keyValue.value.value(),
-              keyValue.value.timestamp()))
+                                  keyValue.value.timestamp()))
```

Comment:
This is not for this PR, but just a thought: could we save on two iterations from the store query result here, by having a delegating iterator of `Iterator<Row>` whose implementation calls the underlying `Iterator<GenericKey, ValueAndTimestamp<GenericRow>>`? I'm asking only because for large returned iterable sets, `Streams.stream()` will iterate and copy-paste into another set and the returns an iterator, which may take non-negligible time.

Reply:
Yes, that makes total sense. I will open a ticket to implement it in a follow up PR.

Reply:
Is this right? It looks to me like the `Stream.stream` and the subsequent `Stream#iterator` are both lazy.